Add asset optimization reporting module

cwasicki opened the pull request:
This pull request introduces a new module, `asset_optimization.py`, for reporting and visualizing microgrid data. The module provides functionality to fetch, process, and plot various aspects of microgrid data, such as power flow, energy trade, battery power, and monthly aggregates.

### New Features for Data Fetching and Processing:
* Added the `fetch_data` function to retrieve and process microgrid data, including support for state of charge (SOC) and component-specific data. It also handles data validation and preprocessing for visualization.

### Visualization Enhancements:
* Implemented `plot_power_flow`, `plot_energy_trade`, and `plot_power_flow_trade` functions to visualize power flow and energy trade data, supporting detailed insights into microgrid operations.
* Added `plot_battery_power` to visualize battery power and SOC, with enhanced labeling and grid customization for clarity.
* Introduced `plot_monthly` for generating monthly aggregate energy data, including positive (consumption) and negative (production) energy flows, with bar chart visualizations.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+        fetch_soc: Whether to fetch state of charge (SOC) data.
+
+    Returns:
+        pd.DataFrame: DataFrame containing the processed data.
```
[nitpick] `pd.DataFrame` is redundant here as it has already been defined before.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+            f"No data found for microgrid {mid} between {start_time} and {end_time}"
+        )
+
+    print(f"Received {df.shape[0]} rows and {df.shape[1]} columns")
```
Why not adding a logger instead of printing?

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+        df = pd.concat([df, soc_df.rename(columns={"battery": "soc"})[["soc"]]], axis=1)
+
+    df["soc"] = df.get("soc", np.nan)
+    # For later vizualization we default to zero
```
typo. "visualization"

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+
+    # Determine consumption if not present
+    if "consumption" not in df.columns:
+        if any(
+            ct not in ["grid", "pv", "battery", "chp", "consumption"]
+            for ct in df.columns
+        ):
+            raise ValueError(
+                "Consumption not found in data and unexpected component types present."
+            )
```
I think this block should be moved outside and before the `if "consumption" not in df.columns:`. The `ValueError` message should be updated accordingly.

cwasicki replied:
The motivation for this is that consumption is derived from other components and it should ensure that all existing ones are accounted for. If we add other components in future this should fail. Other parts of the module doesn't require to account for all components, in fact they are designed to work on a user-defined subset.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+
+
+def plot_power_flow(df: pd.DataFrame, ax: Axes | None = None) -> None:
+    """Plot the power flow of the microgrid."""
```
Docstring could be more descriptive.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+# License: MIT
+# Copyright © 2025 Frequenz Energy-as-a-Service GmbH
+
+"""Reporting for asset optimization."""
```
A few more details in the module docstring would be helpful I think.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+
+def plot_power_flow(df: pd.DataFrame, ax: Axes | None = None) -> None:
+    """Plot the power flow of the microgrid."""
+    d = -df.copy()
```
What's the thinking behind negating all the values here? It makes the rest of the function a bit confusing to me.

cwasicki replied:
Will add an explanation in the docs

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+
+
+def plot_energy_trade(df: pd.DataFrame, ax: Axes | None = None) -> None:
+    """Plot the energy trade of the microgrid."""
```
Docstring could be more descriptive.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+
+def plot_energy_trade(df: pd.DataFrame, ax: Axes | None = None) -> None:
+    """Plot the energy trade of the microgrid."""
+    d = -df.copy()
```
same comment as before.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+    prod = chp + (d["pv"].clip(lower=0) if has_pv else 0)
+
+    if ax is None:
+        fig, ax = plt.subplots(figsize=(30, 10), sharex=True)
```
`sharex` has no effect here right? Only 1 axis is created.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+    prod = chp + (d["pv"].clip(lower=0) if has_pv else 0)
+    trade -= prod
+
+    g = trade.resample("15min").mean() / 4
```
Consider renaming `g` to something more descriptive, like `trade_kwh` or `energy_trade_15min`, to improve readability.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+    g = trade.resample("15min").mean() / 4
+
+    if ax is None:
+        fig, ax = plt.subplots(figsize=(30, 10), sharex=True)
```
Same comment about `sharex` as before.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+    # Plot Battery SOC
+    twin_ax = ax1.twinx()
+    assert df["soc"].ndim == 1, "SOC data should be 1D"
+    soc = df["soc"]  # .iloc[:, 0] if df["soc"].ndim > 1 else df["soc"]
```
The inline comment next to `soc = df['soc']` seems like a leftover or debugging note. It might be worth removing or rephrasing for clarity.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+def plot_monthly(df: pd.DataFrame) -> pd.DataFrame:
+    """Plot monthly aggregate data."""
+    months = df.resample("1MS").sum()
+    resolution = (df.index[1] - df.index[0]).total_seconds()
```
This assumes regular sampling in the DataFrame. Is it guaranteed? If not, consider adding a check and fallback to handle irregular intervals.

cwasicki replied:
It is guaranteed when fetching data with the function in this module. But you are right, will add a validation.

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
```diff
@@ -0,0 +1,330 @@
+        months.index = pd.to_datetime(months.index)
+    months.index = pd.Index(months.index.date)
+    pos, neg = (
+        months[[c for c in months.columns if "_pos" in c]],
+        months[[c for c in months.columns if "_neg" in c]],
```
perhaps a short inline comment to explain where `_pos` and `_neg` are coming from could be useful?

cyiallou commented on src/frequenz/lib/notebooks/reporting/asset_optimization.py:
I wonder if the module could benefit from some refactoring — possibly wrapping the logic into a class to group shared functionality, reduce repetition, and make future extensions easier. Unless the intention is to keep this as a lightweight, script-style utility, in which case the current functional structure might be more appropriate.

General comments:
- Without seeing the plots, I can’t comment directly on the choice of colours and styles. However, I’m confident that the resulting visualisations are clear and well-designed.
- Some function docstrings could be more descriptive (Missing Args, Returns, Raises sections).

cwasicki replied:
It's indeed intended to keep it lightweight as possible, to re-use functions on simple data frames. Also for the visualization functions, the output is not going to be consumed by something else.

cyiallou reviewed (changes requested):
Thanks for the solid work on this module. The structure and logic are generally clear. I have left a few comments, mostly around small improvements to clarity, docstrings, naming, and some safety checks. I also suggested a possible refactor toward a class-based design for better organization, though that may depend on intended usage. I recommend addressing these comments before merging.

cwasicki committed:
Add asset optimization reporting package

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

cwasicki committed:
Add data fetching for asset optimization

Add helper functions to initialize an microgrid data instance from
environment variables. Those can exist in the environment or optionally
be specified in an env file that can be passed to the function.

In addition to that a function for fetching and preparing the data as
required for the visualization functions is added.

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

cwasicki committed:
Add matplotlib plotting for asset optimization

A function that can be used to visualize the power flow between
batteries, power generating and consuming assets in a stacked view is
provided. A second function can be used to plot the excess and deficit
net power, battery charge and battery SoC. Finally functionality to plot
monthly aggregates is added.

All functions in this module are based on matplotlib.

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

cwasicki committed:
Add asset optimization reporting notebook

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

cwasicki committed:
Update release notes

Signed-off-by: cwasicki <126617870+cwasicki@users.noreply.github.com>

cwasicki commented:
@Mohammad-Tayyab-Frequenz  There is an issue with the plotly version of the graphs, maybe I'll remove those from this PR.

cwasicki commented:
Removed the plotly module, will add it in a separate PR. 

@cyiallou @Mohammad-Tayyab-Frequenz  Could one of you have a look please.

cyiallou reviewed (approved):
Looks good to me
